src/CONST.js

```diff
@@ -806,6 +806,7 @@ const CONST = {
 
         // eslint-disable-next-line max-len, no-misleading-character-class
         EMOJIS: /[\p{Extended_Pictographic}\u200d\u{1f1e6}-\u{1f1ff}\u{1f3fb}-\u{1f3ff}\u{e0020}-\u{e007f}\u20E3\uFE0F]|[#*0-9]\uFE0F?\u20E3/gu,
+        EMOJI_SURROGATE: /(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])/,
```

Comment:
Can you explain what this regex represents? I was not able to figure it out

Reply:
From the algorithm, we were expected to capture emojis in a text, though while doing this we need to be careful when we capture emojis that are surrogate pairs.
Surrogate pair Emojis are emojis that can be made up of multiple values that are overlayed on each other to get the final displayed representation.
An example of this is the cloud face emoji made up of 2 different emoji (face emoji and cloud emoji) combined together.
If we split the cloud face emoji we would end up with both separate emojis, so from our algorithm, we need to identify these sorts of surrogate pairs and have them combined back together to maintain the original emoji

![Screenshot 2023-03-03 at 6 58 45 PM](https://user-images.githubusercontent.com/33870502/222793730-a5901f6c-b7a2-4236-b923-421e668ab92e.png)

[Read more on surrogate pair characters](https://secure.jadeworld.com/JADETech/JADE2020/OnlineDocumentation/content/resources/devref/ch1languageref/unicodesurrogatepaircharacters.htm#:~:text=Emojis%20can%20be%20made%20up,each%20part%20of%20the%20emoji)

Reply:
Awesome, thanks for that explanation.

Maybe what we should do then is:

1. Update the emoji regex to capture chunks of 1 or more emojis
1. For each chunk, use the emoji surrogate regex to capture and "squash" surrogate pairs

Reply:
Also we should include some comments to explain what this regex is about

Reply:
@roryabraham 
1. Update the emoji regex to capture chunks of 1 or more emojis
   -   How do I do this from your implementation?

Reply:
> > the d flag works for the web but not on android and ios platforms as they both crash
> 
> Have logs for this crash? Could this be because the `d` flag is not implemented in the Hermes JS engine, or something else?
> 
> **Edit:** Yep, it seems that the `hasIndices` flag is not implemented in Hermes: https://github.com/facebook/hermes/blob/417190297242f14fed540378390d28926c2a7f7e/include/hermes/Regex/RegexTypes.h#L213-L343. Shoot ... did not see that coming.
> 
> **Edit 2:** Created [facebook/hermes#932](https://github.com/facebook/hermes/issues/932), but we need to decide if we should HOLD this PR on that or not

@roryabraham @parasharrajat 

> > the d flag works for the web but not on android and ios platforms as they both crash
> 
> Have logs for this crash? Could this be because the `d` flag is not implemented in the Hermes JS engine, or something else?
> 
> **Edit:** Yep, it seems that the `hasIndices` flag is not implemented in Hermes: https://github.com/facebook/hermes/blob/417190297242f14fed540378390d28926c2a7f7e/include/hermes/Regex/RegexTypes.h#L213-L343. Shoot ... did not see that coming.
> 
> **Edit 2:** Created [facebook/hermes#932](https://github.com/facebook/hermes/issues/932), but we need to decide if we should HOLD this PR on that or not

@roryabraham @parasharrajat this would prolong this task by extra few days should we try to fix this regex issue.
I would suggest we go with the former implementation which worked perfectly avoiding the need for this regex `d` fix also if it helps I would put a comment on it explaining every step of the algorithm and also remove all the other functions that aren't needed.

```js
/**
 * Get all the emojis in the message
 * @param {String} text
 * @returns {Array}
 */
function getAllEmojiFromText(text) {
    // return empty array when no text is passed
    if (!text) {
        return [];
    }

    const result = [];

    // Unicode Character 'ZERO WIDTH JOINER' (U+200D) usually used to join surrogate pair together without braking the emoji
    const emptySpaceChecker = '\u200D'; // https://codepoints.net/U+200D?lang=en
    const emojiCharacterSpaceReplacer = '~'; // unique identifier to replace 'ZERO WIDTH JOINER' in our first string split
    const splittedMessage = text.split('');
    const convertedSplittedMessage = [];

    for (let i = 0; i < splittedMessage.length; i++) {
        const char = splittedMessage[i];

        // replace every 'ZERO WIDTH JOINER' with '~' so we can easily identify where a surrogate pair starts and concatenate them together
        convertedSplittedMessage.push((char === emptySpaceChecker) ? emojiCharacterSpaceReplacer : char);
    }

    let wordHolder = ''; // word counter
    let emojiHolder = ''; // emoji counter

    const setWord = word => result.push({text: word, isEmoji: false});

    // replace all '~' back with a 'ZERO WIDTH JOINER', this helps maintain surrogate pair just as there where before splitting
    const setEmoji = word => result.push({text: Str.replaceAll(word, emojiCharacterSpaceReplacer, emptySpaceChecker), isEmoji: true});

    _.forEach(convertedSplittedMessage, (word, index) => {
        if (CONST.REGEX.EMOJI_SURROGATE.test(word) || word === emojiCharacterSpaceReplacer) {
            setWord(wordHolder);
            wordHolder = '';
            emojiHolder += word;
        } else {
            setEmoji(emojiHolder);
            emojiHolder = '';
            wordHolder += word;
        }

        if (index === convertedSplittedMessage.length - 1) {
            setEmoji(emojiHolder);
            setWord(wordHolder);
        }
    });

    // remove none next characters like '' only return where text is a word or white space ' '
    return _.filter(result, res => res.text);
}


/**
 * Validates that this message contains has emojis
 *
 * @param {String} message
 * @returns {Boolean}
 */
function hasEmojis(message) {
    const splitText = getAllEmojiFromText(message);
    return _.find(splitText, chunk => chunk.isEmoji) !== undefined;
}

/**
 * Validates that this message contains only emojis
 *
 * @param {String} message
 * @returns {Boolean}
 */
function containsOnlyEmojis(message) {
    const splitText = getAllEmojiFromText(message);
    return _.every(splitText, chunk => chunk.isEmoji);
}
```